File: backend/src/algorithms/tokenBucket.ts
import { RateLimitResult } from '../types/rateLimiter.js';

export interface TokenBucketParams {
  tokens: number;
  lastRefillAt: Date | number;
  burstSize: number;
  requestsPerSecond: number;
}

export interface TokenBucketEvaluation extends RateLimitResult {
  newTokens: number;
  newLastRefillAt: Date;
}

/**
 * Pure Token Bucket algorithm implementation.
 *
 * Mathematical properties:
 * - Refills continuously at `requestsPerSecond` rate: tokensToAdd = deltaSeconds * requestsPerSecond
 * - Caps at `burstSize` maximum capacity
 * - Consumes 1 (or `cost`) token if available, returning ALLOW
 * - Denies request if insufficient tokens, returning DENY with retryAfter
 */
export function evaluateTokenBucket(
  params: TokenBucketParams,
  now: Date | number = Date.now(),
  cost: number = 1
): TokenBucketEvaluation {
  const { burstSize, requestsPerSecond } = params;

  if (burstSize <= 0) {
    throw new Error('burstSize must be greater than 0');
  }
  if (requestsPerSecond <= 0) {
    throw new Error('requestsPerSecond must be greater than 0');
  }
  if (cost <= 0) {
    throw new Error('cost must be greater than 0');
  }

  const nowMs = typeof now === 'number' ? now : now.getTime();
  const lastRefillMs =
    typeof params.lastRefillAt === 'number'
      ? params.lastRefillAt
      : params.lastRefillAt.getTime();

  // Elapsed time in seconds (non-negative)
  const elapsedSeconds = Math.max(0, (nowMs - lastRefillMs) / 1000);

  // Tokens added since last refill
  const tokensToAdd = elapsedSeconds * requestsPerSecond;

  // Refill tokens, bounded by burstSize
  const currentTokens = Math.min(burstSize, params.tokens + tokensToAdd);

  const nowSec = Math.floor(nowMs / 1000);

  if (currentTokens >= cost) {
    const newTokens = currentTokens - cost;
    const remaining = Math.max(0, Math.floor(newTokens));

    // Reset is the timestamp when the bucket will be completely full again
    const secondsToFull = requestsPerSecond > 0 ? (burstSize - newTokens) / requestsPerSecond : 0;
    const reset = Math.ceil((nowMs / 1000) + secondsToFull);

    return {
      decision: 'ALLOW',
      algorithm: 'token_bucket',
      remaining,
      limit: Math.floor(burstSize),
      reset: Math.max(nowSec, reset),
      newTokens,
      newLastRefillAt: new Date(nowMs),
    };
  } else {
    // DENIED: tokens remain unchanged (or updated to current refilled amount)
    const remaining = Math.max(0, Math.floor(currentTokens));

    // Seconds required until at least 1 token is available
    const tokensNeeded = cost - currentTokens;
    const secondsToWait = requestsPerSecond > 0 ? Math.max(1, Math.ceil(tokensNeeded / requestsPerSecond)) : 1;

    // Reset timestamp when bucket will reach 1 token (or full)
    const secondsToFull = requestsPerSecond > 0 ? (burstSize - currentTokens) / requestsPerSecond : 0;
    const reset = Math.ceil((nowMs / 1000) + secondsToFull);

    return {
      decision: 'DENY',
      algorithm: 'token_bucket',
      remaining,
      limit: Math.floor(burstSize),
      reset: Math.max(nowSec, reset),
      retryAfter: secondsToWait,
      newTokens: currentTokens,
      newLastRefillAt: new Date(nowMs),
    };
  }
}

File: backend/src/repositories/lua/tokenBucket.lua
-- Redis Lua Script: Atomic Token Bucket Evaluation
-- KEYS[1]: client:{clientKey}:bucket
-- ARGV[1]: burstSize (capacity)
-- ARGV[2]: requestsPerSecond (refill rate)
-- ARGV[3]: nowMs (current Unix epoch in milliseconds)
-- ARGV[4]: cost (tokens to consume, default 1)

local bucketKey = KEYS[1]
local burstSize = tonumber(ARGV[1])
local requestsPerSecond = tonumber(ARGV[2])
local nowMs = tonumber(ARGV[3])
local cost = tonumber(ARGV[4] or "1")

local state = redis.call('HMGET', bucketKey, 'tokens', 'last_refill_at')
local currentTokens = nil
local lastRefillMs = nil

if state and state[1] and state[1] ~= false and tostring(state[1]) ~= "" then
  currentTokens = tonumber(state[1])
  lastRefillMs = tonumber(state[2])
end

if not currentTokens then
  currentTokens = burstSize
end
if not lastRefillMs then
  lastRefillMs = nowMs
end

local elapsedSeconds = math.max(0, (nowMs - lastRefillMs) / 1000.0)
local tokensToAdd = elapsedSeconds * requestsPerSecond
local refilledTokens = math.min(burstSize, currentTokens + tokensToAdd)

local nowSec = math.floor(nowMs / 1000.0)
local limit = math.floor(burstSize)

if refilledTokens >= cost then
  local newTokens = refilledTokens - cost
  redis.call('HSET', bucketKey, 'tokens', tostring(newTokens), 'last_refill_at', tostring(nowMs))
  redis.call('EXPIRE', bucketKey, 604800)

  local remaining = math.max(0, math.floor(newTokens))
  local secondsToFull = requestsPerSecond > 0 and ((burstSize - newTokens) / requestsPerSecond) or 0
  local reset = math.max(nowSec, math.ceil((nowMs / 1000.0) + secondsToFull))

  -- Return { is_allowed (1/0), remaining, limit, reset, retry_after }
  return { 1, remaining, limit, reset, 0 }
else
  redis.call('HSET', bucketKey, 'tokens', tostring(refilledTokens), 'last_refill_at', tostring(nowMs))
  redis.call('EXPIRE', bucketKey, 604800)

  local tokensNeeded = cost - refilledTokens
  local retryAfter = requestsPerSecond > 0 and math.max(1, math.ceil(tokensNeeded / requestsPerSecond)) or 1
  local secondsToFull = requestsPerSecond > 0 and ((burstSize - refilledTokens) / requestsPerSecond) or 0
  local reset = math.max(nowSec, math.ceil((nowMs / 1000.0) + secondsToFull))

  return { 0, 0, limit, reset, retryAfter }
end
